File: src/main/java/org/stt/update/VersionComparator.java
package org.stt.update;

import java.io.Serializable;
import java.util.Arrays;
import java.util.Comparator;

public class VersionComparator implements Comparator<String>, Serializable {
    @Override
    public int compare(String a, String b) {
        int[] aVersionParts = Arrays.stream(a.split("\\.")).mapToInt(this::asVersionPart).toArray();
        int[] bVersionParts = Arrays.stream(b.split("\\.")).mapToInt(this::asVersionPart).toArray();
        for (int i = 0; i < Math.min(aVersionParts.length, bVersionParts.length); i++) {
            if (aVersionParts[i] < bVersionParts[i]) {
                return -1;
            } else if (aVersionParts[i] > bVersionParts[i]) {
                return 1;
            }
        }
        return Integer.compare(aVersionParts.length, bVersionParts.length);
    }

    private Integer asVersionPart(String s) {
        try {
            return Integer.parseInt(s);
        } catch (NumberFormatException e) {
            return 0;
        }
    }
}

File: src/main/kotlin/org/stt/update/VersionComparator.kt
package org.stt.update

import java.io.Serializable
import java.util.*

class VersionComparator : Comparator<String>, Serializable {
    override fun compare(a: String, b: String): Int {
        val aVersionParts = a.split("[.-]".toRegex()).map { this.asVersionPart(it) }
        val bVersionParts = b.split("[.-]".toRegex()).map { this.asVersionPart(it) }
        for (i in 0 until Math.min(aVersionParts.size, bVersionParts.size)) {
            if (aVersionParts[i] < bVersionParts[i]) {
                return -1
            } else if (aVersionParts[i] > bVersionParts[i]) {
                return 1
            }
        }
        return Integer.compare(aVersionParts.size, bVersionParts.size)
    }

    private fun asVersionPart(s: String): Int {
        try {
            return Integer.parseInt(s)
        } catch (e: NumberFormatException) {
            return 0
        }

    }
}
